### PR TITLE
detect invalid path expression (fix #862)

### LINKED ISSUE #862: Complex assignments bug
Looks like a bug to me

This works:
jq '(.posts[] | select(.author == "stedolan") | .comments) |= . + ["terrible."]'<<<'{ "posts": [ { "comments": ["legend"], "author": "Bradman" }, { "comments": ["rad"], "author": "Tesla" }, { "comments": ["okay"], "author": "stedolan" } ] }'

This clearly doesn't work:
jq '(.posts | map(select(.author == "stedolan")) | .[].comments) |= . + ["terrible."]'<<<'{ "posts": [ { "comments": ["legend"], "author": "Bradman" }, { "comments": ["rad"], "author": "Tesla" }, { "comments": ["okay"], "author": "stedolan" } ] }'

Tested on an up to date Amazon Linux distro from a git clone of the master branch of jq jq-1.5rc1-157-gf5a7bba 2015/07/22

### PR DESCRIPTION
**Pull request by dtolnay**

### CODE REVIEW

File: execute.c [hunk 1]
```diff
@@ -536,17 +555,30 @@ jv jq_next(jq_state *jq) {
       stack_save(jq, pc - 1, stack_get_pos(jq));
 
       stack_push(jq, jv_number(jq->subexp_nest));
+      stack_push(jq, jv_copy(jq->value_at_path));
```
Comment (nicowilliams) [inline execute.c]: No need to jv_copy(jq->value_at_path) just to jv_free() it immediately below.

File: execute.c [hunk 2]
```diff
@@ -536,17 +555,30 @@ jv jq_next(jq_state *jq) {
...
       stack_push(jq, v);
```
Comment (nicowilliams) [inline execute.c]: This jv_free()s v, but you refer to v below!

File: execute.c [hunk 3]
```diff
@@ -536,17 +555,30 @@ jv jq_next(jq_state *jq) {
...
 
       jq->path = jv_array();
+      jv_free(jq->value_at_path);
+      jq->value_at_path = jv_copy(v);
```
Comment (nicowilliams) [inline execute.c]: See above.

Comment (nicowilliams) [inline execute.c]: Er,  is v really the value at path?

**Comment by nicowilliams** [comment]
I won't be around to respond till late tonight, FYI.

File: execute.c (hunk 2 above)
Comment (dtolnay) [inline execute.c]: No it doesn't, but I will move jv_copy from line 563 to here. I guess the reason I had it below was I conceptualized v as "a value on the stack," even though we temporarily pop it off to insert things below it. So the stack is holding a reference to v. On line 563, also putting a reference to v into the jq_state requires a jv_copy.

File: execute.c (hunk 1 above)
Comment (dtolnay) [inline execute.c]: Good catch.

File: execute.c (hunk 3 above)
Comment (dtolnay) [inline execute.c]: Yes. The path is empty at this point, and v is the "." object on top of the stack as we start building the path. The next INDEX operation must index into v.

**Comment by dtolnay** [comment]
I pushed two new commits, one addressing comments and one fixing a bug. I will squash these before merging.

**Comment by dtolnay** [comment]
Squashed and merged 16e8d0b.

**Pull request closed without merging**